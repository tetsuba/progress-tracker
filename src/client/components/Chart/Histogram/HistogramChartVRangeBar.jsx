import React from 'react'

type Props = {
  correct?: boolean,
  wrong?: boolean,
  height: number,
}

export default function HistogramChartVRangeBar(props: Props) {
  const { correct, height } = props
  const style = {
    width: '100%',
    height: `${height}%`,
    backgroundColor: correct ? '#309c3f' : 'red',
    boxSizing: 'border-box',
    borderTop: 'solid 1px #CCC',
  }

  return <div style={style} />
}
