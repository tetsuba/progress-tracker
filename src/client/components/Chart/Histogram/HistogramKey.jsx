import React from 'react'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSquare } from '@fortawesome/free-solid-svg-icons'

export default function HistogramKey() {
  const styles = {
    textAlign: 'center',
    width: '100%',
  }
  return (
    <div className="HistogramKey" style={styles}>
      <span style={{ color: '#309c3f', marginRight: '20px' }}>
        <FontAwesomeIcon icon={faSquare} /> Correct
      </span>
      <span style={{ color: 'red' }}>
        <FontAwesomeIcon icon={faSquare} /> wrong
      </span>
    </div>
  )
}
