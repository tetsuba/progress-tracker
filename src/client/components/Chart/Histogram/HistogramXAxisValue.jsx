import React from 'react'

type Props = {
  letter: string,
}

export default function HistogramXAxisValue(props: Props) {
  const { letter } = props
  const styles = {
    width: '3%',
    textAlign: 'center',
  }

  return <div style={styles}>{letter}</div>
}
