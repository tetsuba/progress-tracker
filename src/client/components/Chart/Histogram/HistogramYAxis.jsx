import React from 'react'
import HistogramYAxisValue from './HistogramYAxisValue'

type Props = {
  data: any,
}

export default function HistogramYAxis(props: Props) {
  const { data } = props
  const YAxis = Array(data.a.total).fill(0)
  const styles = {
    width: '1%',
    paddingBottom: '42px',
  }

  return (
    <div className="HistogramYAxis" style={styles}>
      <div
        style={{
          position: 'absolute',
          transform: 'rotate(-90deg)',
          transformOrigin: '0 center',
          top: '60%',
          left: '0',
          color: '#666',
          fontSize: '14px',
        }}
      >
        Number of assessments
      </div>
      {YAxis.map((_, index) => (
        <HistogramYAxisValue key={`${index}-yAxis`} data={data} index={index} />
      )).reverse()}
    </div>
  )
}
