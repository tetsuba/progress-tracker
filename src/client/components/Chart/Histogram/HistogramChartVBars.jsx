import React from 'react'
import HistogramChartVRangeBar from './HistogramChartVRangeBar'
import { OverlayTrigger, Tooltip } from 'react-bootstrap'

type Props = {
  data: {
    history: [number],
    correct: number,
    wrong: number,
    total: number,
  },
}

export default function HistogramChartVBars(props: Props) {
  const { data } = props

  const style = {
    width: '100%',
    height: '100%',
    backgroundColor: '#CCC',
    padding: '1px',
    alignItems: 'flex-end',
    display: 'flex',
    flexDirection: 'column',
  }

  const height = 100 / data.total
  const placement = 'top'

  return (
    <div
      style={{
        display: 'flex',
        flexDirection: 'column',
        width: '3%',
      }}
    >
      <OverlayTrigger
        key={placement}
        placement={placement}
        overlay={
          <Tooltip id={`tooltip-${placement}`}>
            <strong>
              {data.correct} Correct {data.wrong} Wrong
            </strong>
            .
          </Tooltip>
        }
      >
        <div style={style}>
          {data.history
            .map((value, index) => {
              switch (value) {
                case 1:
                  return (
                    <HistogramChartVRangeBar
                      key={`correct-${index}`}
                      height={height}
                      correct
                    />
                  )
                case 2:
                  return (
                    <HistogramChartVRangeBar
                      key={`wrong-${index}`}
                      height={height}
                      wrong
                    />
                  )
                default:
                  return null
              }
            })
            .reverse()}
        </div>
      </OverlayTrigger>
    </div>
  )
}
