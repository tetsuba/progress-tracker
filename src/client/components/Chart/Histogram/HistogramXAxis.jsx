import React from 'react'
import HistogramXAxisValue from './HistogramXAxisValue'

type Props = {
  data: any,
}

export default function HistogramXAxis(props: Props) {
  const { data } = props
  const values = Object.keys(data)
  const styles = {
    width: '100%',
    display: 'flex',
    flexDirection: 'row',
    flexFlow: 'row nowrap',
    justifyContent: 'space-between',
    padding: '0 0 8px 8px',
  }
  return (
    <div className="HistogramXAxis" style={styles}>
      {values.map((value) => (
        <HistogramXAxisValue key={`${value}-xAxis`} letter={value} />
      ))}
    </div>
  )
}
