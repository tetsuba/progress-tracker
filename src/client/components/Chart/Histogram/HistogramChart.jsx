import React from 'react'
import HistogramChartVBars from './HistogramChartVBars'
import type { HistogramDataType } from './Histogram'

type Props = {
  data: HistogramDataType,
}

export default function HistogramChart(props: Props) {
  const { data } = props
  const styles = {
    width: '100%',
    height: '400px',
    display: 'flex',
    flexDirection: 'row',
    flexFlow: 'row nowrap',
    justifyContent: 'space-between',
    borderBottom: 'solid 1px #666',
    borderLeft: 'solid 1px #666',
    padding: '0 0 8px 8px',
  }

  const bars: Array<[string, any]> = Object.entries(data)

  return (
    <div style={styles}>
      {bars.map(([key, value]) => (
        <HistogramChartVBars key={`${key}-bar`} data={value} />
      ))}
    </div>
  )
}
