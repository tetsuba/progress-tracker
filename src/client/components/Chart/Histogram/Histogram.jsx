import React from 'react'
import HistogramYAxis from './HistogramYAxis'
import HistogramXAxis from './HistogramXAxis'
import HistogramChart from './HistogramChart'
import HistogramKey from './HistogramKey'

export type HistogramDataType = {
  [string]: {
    history: [number],
    correct: number,
    wrong: number,
    total: number,
  },
}

type Props = {
  data: HistogramDataType,
}

export default function Histogram(props: Props) {
  const { data } = props

  const styles = {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
  }

  return (
    <div className="Histogram" style={styles}>
      <HistogramYAxis data={data} />
      <div style={{ width: '99%' }}>
        <HistogramChart data={data} />
        <HistogramXAxis data={data} />
      </div>
      <HistogramKey />
    </div>
  )
}
