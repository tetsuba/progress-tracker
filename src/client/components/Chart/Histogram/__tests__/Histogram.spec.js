import { testRenderer } from '../../../../../test/testHelper'
import Histogram from '../Histogram'

describe('<Histogram>', () => {
  describe('@Render', () => {
    it('Should render', () => {
      const props = {
        data: {},
      }
      const wrapper = testRenderer(Histogram, props)
      expect(wrapper).toMatchSnapshot()
    })
  })
})
