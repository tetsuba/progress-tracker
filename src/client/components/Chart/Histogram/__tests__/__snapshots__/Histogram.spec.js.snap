// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Histogram> @Render Should render 1`] = `
<div
  className="Histogram"
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
>
  <HistogramYAxis
    data={Object {}}
  />
  <div
    style={
      Object {
        "width": "99%",
      }
    }
  >
    <HistogramChart
      data={Object {}}
    />
    <HistogramXAxis
      data={Object {}}
    />
  </div>
  <HistogramKey />
</div>
`;
